File: Problem1_Mjolnir/1_chatgpt_sol1.java
import java.io.*;
import java.util.*;

class Main {
    static final int INF = 1_000_000_000;

    public static void main(String[] args) throws Exception {
        FastScanner fs = new FastScanner(System.in);
        StringBuilder result = new StringBuilder();

        int t = fs.nextInt();

        while (t-- > 0) {
            int n = fs.nextInt();
            long r = fs.nextLong();

            long sCap = fs.nextLong();
            long wCap = fs.nextLong();

            int[] impact = new int[n];
            int[] cost = new int[n];

            long totalResistance = r;

            for (int i = 0; i < n; i++) {
                int s = fs.nextInt();
                int w = fs.nextInt();
                int c = fs.nextInt();

                impact[i] = Math.max(0, s - w);
                cost[i] = c;

                totalResistance += impact[i];
            }

            long maximumResistance = sCap + wCap;

            if (totalResistance <= maximumResistance) {
                result.append(0).append('\n');
                continue;
            }

            int required = (int) (totalResistance - maximumResistance);

            int[] dp = new int[required + 1];
            Arrays.fill(dp, INF);
            dp[0] = 0;

            for (int i = 0; i < n; i++) {
                for (int removed = required; removed >= 0; removed--) {
                    if (dp[removed] == INF) {
                        continue;
                    }

                    int newRemoved = Math.min(
                            required,
                            removed + impact[i]);

                    dp[newRemoved] = Math.min(
                            dp[newRemoved],
                            dp[removed] + cost[i]);
                }
            }

            result.append(dp[required] == INF ? -1 : dp[required])
                    .append('\n');
        }

        System.out.print(result);
    }

    static class FastScanner {
        private final InputStream input;
        private final byte[] buffer = new byte[1 << 16];
        private int pointer = 0;
        private int length = 0;

        FastScanner(InputStream input) {
            this.input = input;
        }

        private int read() throws IOException {
            if (pointer >= length) {
                length = input.read(buffer);
                pointer = 0;

                if (length == -1) {
                    return -1;
                }
            }

            return buffer[pointer++];
        }

        long nextLong() throws IOException {
            int ch;

            do {
                ch = read();
            } while (ch <= ' ');

            long value = 0;

            while (ch > ' ') {
                value = value * 10 + (ch - '0');
                ch = read();
            }

            return value;
        }

        int nextInt() throws IOException {
            return (int) nextLong();
        }
    }
}

File: Problem1_Mjolnir/1_chatgpt_sol1.py
import sys

input = sys.stdin.readline

t = int(input())

for _ in range(t):
    n, r = map(int, input().split())
    scap, wcap = map(int, input().split())

    people = []
    total = r

    for _ in range(n):
        s, w, c = map(int, input().split())
        impact = max(0, s - w)
        total += impact
        people.append((impact, c))

    allowed = scap + wcap

    if total <= allowed:
        print(0)
        continue

    required = total - allowed
    INF = 10**18

    dp = [INF] * (required + 1)
    dp[0] = 0

    for impact, cost in people:
        for current in range(required, -1, -1):
            if dp[current] == INF:
                continue

            new_amount = min(required, current + impact)
            candidate = dp[current] + cost

            if candidate < dp[new_amount]:
                dp[new_amount] = candidate

    print(-1 if dp[required] == INF else dp[required])
